File: solutions/5. Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        sr=''
        for i in range(0,len(s)):
            sr=sr+'|'+s[i]
        sr=sr+'|'
        
        center=0
        radius=0
        pa=[]
        
        while center<len(sr):
            while center-(radius+1)>=0 and center+(radius+1)<len(sr) and sr[center-(radius+1)]==sr[center+(radius+1)]:
                radius+=1
            pa.append(radius)
            
            
            oldcenter=center
            oldradius=radius
            radius=0
            center=oldcenter+1
            while center<=oldcenter+oldradius:
                mirrorcenter=oldcenter-(center-oldcenter)
                maxradius=oldcenter+oldradius-center
                if pa[mirrorcenter]<maxradius:
                    pa.append(pa[mirrorcenter])
                    center=center+1
                elif pa[mirrorcenter]>maxradius:
                    pa.append(maxradius)
                    center=center+1
                else:
                    radius=maxradius
                    break
                
        pali_len=max(pa)
        pali_center=pa.index(max(pa))
        #print(pa)
        #print(pali_len)
        #print(pali_center)
        #print(int (pali_center-int((pali_len-1)/2)))
        return sr[(pali_center-pali_len):(pali_center+pali_len+1)].replace('|','')
                
                
            
        
            
                
        

File: solutions/5. Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String s) {
        //if (s.length()<=1) {return s;}
        int end=0; int start=0;
        for (int i=0; i<s.length();i++) {
            int n_1=palindrome(s,i,i); 
            int n_2=palindrome(s,i,i+1);
            int n_l=Math.max(n_1,n_2);
            if (n_l>end-start) {
                start=i-(n_l-1)/2;
                end=i+(n_l)/2;
                }
        }
        return s.substring(start,end+1);
    }
        
        public int palindrome(String s, int l, int r) {
            
            while (l>=0 && r<=s.length()-1 && s.charAt(l)==s.charAt(r)) {
                l=l-1;
                r=r+1;
            }
             
             return r-l-1; 
                   }
}
